pail/util.go

```diff
@@ -0,0 +1,69 @@
+package pail
+
+import (
+	"context"
+	"crypto/md5"
+	"crypto/sha1"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+func checksum(hash hash.Hash, path string) (string, error) {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return "", errors.Wrapf(err, "file '%s' does not exist", path)
+	}
+
+	f, err := os.Open(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "problem opening '%s'", path)
+	}
+	defer f.Close()
+
+	_, err = io.Copy(hash, f)
+	if err != nil {
+		return "", errors.Wrap(err, "problem hashing contents")
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil)), nil
+}
+
+func md5sum(path string) (string, error) {
+	out, err := checksum(md5.New(), path)
+	return out, errors.WithStack(err)
+}
+
+func sha1sum(path string) (string, error) {
+	out, err := checksum(sha1.New(), path)
+	return out, errors.WithStack(err)
+}
+
+func walkLocalTree(ctx context.Context, prefix string) ([]string, error) {
+	var out []string
+	err := filepath.Walk(prefix, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return nil
```

Comment:
Seems like this should return an error.

Reply:
It does, but that doesn't work because listing empty buckets shouldn't be an error, and it's fine to just skip bad files, I think.